backend/postgres/postgres.go

```diff
@@ -0,0 +1,955 @@
+					} else {
+						return err
+					}
+				}
+			}
+
+			eventPayload, err := protojson.Marshal(msg.HistoryEvent)
+			if err != nil {
+				return err
+			}
+
+			sqlInsertArgs = append(sqlInsertArgs, msg.TargetInstanceID, eventPayload, nil)
+		}
+
+		_, err = tx.ExecContext(ctx, insertSql, sqlInsertArgs...)
+		if err != nil {
+			return fmt.Errorf("failed to insert into the NewEvents table: %w", err)
+		}
+	}
+
+	// Delete inbound events
+	dbResult, err := tx.ExecContext(
+		ctx,
+		"DELETE FROM durabletask.NewEvents WHERE InstanceID = $1 AND LockedBy = $2",
+		string(wi.InstanceID),
+		wi.LockedBy,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to delete from NewEvents table: %w", err)
+	}
+
+	rowsAffected, err := dbResult.RowsAffected()
+	if err != nil {
+		return fmt.Errorf("failed get rows affected by delete statement: %w", err)
+	} else if rowsAffected == 0 {
+		return backend.ErrWorkItemLockLost
+	}
+
+	if err != nil {
+		return fmt.Errorf("failed to delete from the NewEvents table: %w", err)
+	}
+
+	if err = tx.Commit(); err != nil {
+		return fmt.Errorf("failed to commit transaction: %w", err)
+	}
+
+	return nil
+}
+
+// CreateOrchestrationInstance implements backend.Backend
+func (be *postgresBackend) CreateOrchestrationInstance(ctx context.Context, e *backend.HistoryEvent) error {
+	if err := be.ensureDB(ctx); err != nil {
+		return err
+	}
+
+	tx, err := be.db.BeginTx(ctx, nil)
+	if err != nil {
+		return fmt.Errorf("failed to start transaction: %w", err)
+	}
+	defer tx.Rollback()
+
+	var instanceID string
+	if err := be.createOrchestrationInstanceInternal(ctx, e, tx, &instanceID); err != nil {
+		return err
+	}
+
+	eventPayload, err := protojson.Marshal(e)
+	if err != nil {
+		return err
+	}
+
+	_, err = tx.ExecContext(
+		ctx,
+		`INSERT INTO durabletask.NewEvents (InstanceID, EventPayload) VALUES ($1, $2)`,
+		instanceID,
+		eventPayload,
+	)
+
+	if err != nil {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		return fmt.Errorf("failed to insert row into [NewEvents] table: %w", err)
+	}
+
+	if err = tx.Commit(); err != nil {
+		return fmt.Errorf("failed to create orchestration: %w", err)
+	}
+
+	return nil
+}
+
+func (be *postgresBackend) createOrchestrationInstanceInternal(ctx context.Context, e *backend.HistoryEvent, tx *sql.Tx, instanceID *string) error {
+	if e == nil {
+		return errors.New("HistoryEvent must be non-nil")
+	} else if e.Timestamp == nil {
+		return errors.New("HistoryEvent must have a non-nil timestamp")
+	}
+
+	startEvent := e.GetExecutionStarted()
+	if startEvent == nil {
+		return errors.New("HistoryEvent must be an ExecutionStartedEvent")
+	}
+
+	// TODO: Support for re-using orchestration instance IDs
+	res, err := tx.ExecContext(
+		ctx,
+		`INSERT INTO durabletask.Instances (
+			Name,
+			Version,
+			InstanceID,
+			ExecutionID,
+			Input,
+			RuntimeStatus,
+			CreatedTime
+		) VALUES ($1, $2, $3, $4, $5, $6, $7)
+		ON CONFLICT (InstanceID) DO NOTHING`,
```

Comment:
Don't ignore conflict errors. Instead, let it error in case of duplicate rows, and then do (this assumes pgx, but similar things can be done with stdlib):

```go
if err != nil {
    var pgErr *pgconn.PgError
	if errors.As(err, &pgErr) && pgErr.Code == pgerrcode.UniqueViolation {
        return backend.ErrDuplicateEvent
    }

    return fmt.Errorf("failed to insert into the Instances table: %w", err)
}
```